pkg/endpoint/policy.go

```diff
@@ -498,14 +512,91 @@ func (e *Endpoint) Regenerate(regenMetadata *regeneration.ExternalRegenerationMe
 				e.getLogger().Debug("regeneration was cancelled")
 			}
 		}
-
-		done <- buildSuccess
-		close(done)
 	}()
 
 	return done
 }
 
+var reasonRegenRetry = "retrying regeneration"
+
+// startAsyncRegenerationFailureHandler launches a goroutine which waits for a build of
+// the Endpoint to fail. If the Endpoint is deleted, the goroutine is terminated.
+// If a build fails, a controller is launched which tries to regenerate the
+// Endpoint until it suceeds. Once the controller succeeds, it will not be
+// ran again unless another build failure occurs.
+func (e *Endpoint) startAsyncRegenerationFailureHandler() {
+	var boff backoff.Exponential
+	var numConsecutiveFailures int
+	go func() {
+		for {
+			select {
+			case <-e.regenFailedChan:
+				e.getLogger().Debug("received signal that regeneration failed")
+				select {
```

Comment:
This seems a bit overkill as the aliveCtx can be canceled at any time. This seems to handle the case where both channels become ready almost at the same time and the order is random, however, if with this code, the aliveCtx can be canceled right after so it seems to handle a super tiny race window that still exists even with this code.

The endpoint delete case is handler later on which will skip the build and return `nil`, thus causing the for loop to cycle and the aliveCtx handler being evaluated again.

Reply:
Agreed, I'll remove this inner `select`.